Parameter setting cmm algo

lomamech opened the pull request:
- Add property average consumption algorith is stock setting
- Update form for manage stock setting
- Add define algorith to use for average monthly consuption to Use

closes #4984

mbayopanda commented on client/src/i18n/en/stock.json:
```diff
@@ -101,6 +101,16 @@
     "MONTHLY_CONSUM"  : "Average Monthly Consumption",
+    "MONTHLY_CONSUMPTION" : {
+      "ALGO_1"          : "Algorithm 1",
+      "ALGO_1_COMMENT"  : "The average monthly consumption is obtained by dividing the quantity consumed during the period defined by the number of days of stock for said period, and by multiplying the result obtained by 30.5",
```
Since you mentioned `monthly consumption` and `30.5` which are related to month, is it possible to change `period defined` by just `month`?
Ex. The average monthly consumption is obtained by dividing the quantity consumed during the month by the number of days of stock of the month, and by multiplying the result obtained by 30.5

lomamech replied:
I had agreed with @mbayopanda, for the calculation of the CMM we do not use only one me but rather the period defined in the stock Setting 
```
month_average_consumption 
````

jniles replied:
We should probably come up with a better name than "algo 1", "algo 2", etc.  Also, I'm not sure we should support anything other than algo 1 and 4.

mbayopanda reviewed (approved):
LGTM 👍

jmcameron commented on client/src/modules/stock/settings/stock-settings.html:
```diff
@@ -49,37 +49,62 @@
           <bh-yes-no-radios label="SETTINGS.ENABLE_DAILY_CONSUMPTION_LABEL"
             value="StockSettingsCtrl.settings.enable_daily_consumption"
             name="enable_daily_consumption"
+            id="enable_daily_consumption"
```
Great idea adding these tags so we can access them easily in the tests!

jmcameron commented on client/src/modules/stock/settings/stock-settings.js:
```diff
@@ -43,6 +43,14 @@ function StockSettingsController(
       .catch(Notify.handleError);
+
+    // load algorithmes for Average Consumption
+    StockSettings.getAverageConsuptionAlgo()
```
I find it hard to keep track of which algorithm is 1, 2, etc.   I suggest appending the algorithm number on these functions:    getAvertageConsomptionAlgo1()     (add the correct number!)
That way is is easier to track what is happening.

lomamech replied:
This function just allows you to obtain the different existing algorithms...
````
// GET '/stock/setting/average_consuption'
// Get the different algorithms for calculating the average monthly consumption
exports.averageConsuptionAlgo = function averageConsuptionAlgo(req, res, next) {
  const sql = `SELECT id, name, label, comment FROM average_consumption_algo;`;

  db.exec(sql)
    .then((rows) => {
      res.status(200).json(rows);
    })
    .catch(next)
    .done();
};

````
![image](https://user-images.githubusercontent.com/8035588/95433083-682c6500-0947-11eb-80e7-0c845ef9282f.png)

jmcameron replied:
Ok.  Thx.

jmcameron commented on test/end-to-end/stock/stock.setting.js:
```diff
@@ -0,0 +1,48 @@
+    await page.setRadio('yes', 'enable_auto_stock_accounting');
+    await page.setRadio('yes', 'enable_supplier_credit');
+    await element(by.id('algo_msh')).click();
+
```
Can we verify that these items have actually changed?  (also in the next 'it').

jniles replied:
Looks like this was addressed.

jmcameron commented on server/models/schema.sql:
```diff
@@ -1849,6 +1849,17 @@ CREATE TABLE IF NOT EXISTS `voucher_item` (
+DROP TABLE IF EXISTS `average_consumption_algo`;
+
+CREATE TABLE `average_consumption_algo` (
+  `id` TINYINT(3) UNSIGNED NOT NULL AUTO_INCREMENT,
```
It is not clear to me why we need a table for the algorithms?   They seem to be something that is strictly in the Javascript code.

lomamech replied:
The reason we used a table is just because we had thought to keep the identifier of the algorithm to be used among the stock settings.

jmcameron replied:
That sounds fine.  Can't we use an enum for the algorithm and add a new field in the stock_settings table?  I'm still not clear why we need a new table for the algorithm info.

lomamech replied:
In this case we could use the component **bhConstants** and modify the property
````
 `average_consumption_algo_id` TINYINT(3) UNSIGNED NOT NULL DEFAULT 3,
````
In VARCHAR with the following values 'algo1', 'algo2', 'algo' and 'algo_msh'

lomamech replied:
I will make this modification

jmcameron reviewed (approved):
Great job!   

I ran this on system ('yarn dev') and it seems to work fine.

The new tests ran fine ('yarn test:ends').
Note that the unrelated distribution keys failed on my system (see Issue https://github.com/IMA-WorldHealth/bhima/issues/5006).  Since it passes on the CI testing, this is nothing for you to worry about.

I'm unclear: I see that we can select the CMM algorithm, etc, but are we actually using that algorithm in the CMM calculations (eg on the Items in Stock page)?

Also, when we start using the selected algorithm in the items in stock page, I suggest adding a tool tip on the "CMM" column header that shows which algorithm is being used for the calculation (when you hover over it).

lomamech committed:
test(Stock Setting)
- Add test to expect the item changed

lomamech committed:
Fix Conflict

lomamech committed:
Fix and resolve conflict

lomamech committed:
Fix and resolve conflict with master

lomamech committed:
Clean and sanitaze code

lomamech committed:
refactor(Stock Setting)
- Use bhConstant instead table for keep the identifier of the algorithme
  to use for calculate Average Consumption

jniles commented on client/src/i18n/fr/stock.json:
```diff
@@ -182,7 +192,7 @@
     "ENABLE_AUTO_STOCK_ACCOUNTING_HELP_TEXT" : "L'activation de cette option va écrire automatiquement et en temps réel dans le journal toutes les transactions liées aux mouvements de stock. Les comptes des inventaires et groupes d'inventaire doivent être bien configurés au préalable",
     "ENABLE_DAILY_CONSUMPTION_LABEL" : "Activer le calcul du CMM sur base des jours de consommation",
-    "ENABLE_DAILY_CONSUMPTION_HELP_TEXT" : "La Cons mation Moyenne Mensuelle (CMM) est calculée en tenant compte des quantités consommées sur le nombre de jours où il y a eu des consommation",
+    "ENABLE_DAILY_CONSUMPTION_HELP_TEXT" : "La Consommation Moyenne Mensuelle (CMM) est calculée en tenant compte des quantités consommées sur le nombre de jours où il y a eu des consommation",
```
:+1:

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -73,4 +73,31 @@ DROP TABLE IF EXISTS `depot_distribution_permission`;
+  (3, 'algo3', 'STOCK.MONTHLY_CONSUMPTION.ALGO_3', 'STOCK.MONTHLY_CONSUMPTION.ALGO_3_COMMENT'),
+  (4, 'algo_msh', 'STOCK.MONTHLY_CONSUMPTION.ALGO_4', 'STOCK.MONTHLY_CONSUMPTION.ALGO_4_COMMENT');
+
+ALTER TABLE `stock_setting` ADD COLUMN `average_consumption_algo_id` TINYINT(3) UNSIGNED NOT NULL DEFAULT 3;
```
Awesome 👍  Thanks for adding these to the migration script.

lomamech replied:
Sorry, I must update the migrate filte

jniles reviewed (approved):
@lomamech great work!  This is a nice and concise PR.

I agree with @jmcameron that we don't really need to store these in the database.  However, I imagine on the long term, we may end up moving a lot of our constants into the database.  You've done a good implementation, if it becomes obvious that they should or should not be in the database, we'll take a further action.

bors r+

lomamech committed:
bug: Fix migrate file remove not necessary script
